### PR TITLE
Illustrate beam_idx

### PR DESCRIPTION
**Pull request by Wovchena**

### CODE REVIEW

File: text_generation/causal_lm/cpp/beam_idx.gif
Comment (ilya-lavrenov) [inline text_generation/causal_lm/cpp/beam_idx.gif:1]: is not 7MBs too much for non-LFS data?

Comment (Wovchena) [reply text_generation/causal_lm/cpp/beam_idx.gif:1]: git lfs is beneficial when large files have different locations in different commits. So checkouts won't copy the files each time. This file is unlikely to move: maybe only when python implementation is added

File: text_generation/causal_lm/cpp/README.md [hunk 1]
```diff
@@ -12,6 +12,8 @@ Hiding KV-cache introduces a peculiarity for beam search algorithm. Beam search
 
 To provide a possibility to implement beam search without accessing model internal state, a stateful LLM converted with `optimum-intel` or [llm_bench](../../../llm_bench/python/) introduces additional 1-dimentional `beam_idx` input. `beam_idx` must contain indices of elements in a batch which are supposed to be selected and evolve during next beam search iteration. Suppose there are two running beams. To proceed generating both beams at the next iteration, `beam_idx` values must be `[0, 1]`, pointing to batch elements `0` and `1`. To drop the last beam and split the other beam in two, `beam_idx` must be set to `[0, 0]`, this results in utilizing only the part of KV cache corresponding to zeroth element in the batch. The process of selecting appropriate entries in cache is called Cache Reorder.
 
+![](beam_idx.gif)
```
Comment (slyalin) [inline text_generation/causal_lm/cpp/README.md:15]: I would recommend to use specific batch size, not n and align description above with the picture.

Comment (slyalin) [reply text_generation/causal_lm/cpp/README.md:15]: Provide a better description for cache modification actions we can achieve by providing different beam_idx: (1) sequence fork when the same kv-cache history is duplicated, (2) when sequence is removed from a batch. Note how we leverage dynamic batch size by mentioning that beam widths can change dynamically. Describe how the first iteration is different from other iterations and it is implemented by beam_idx control.

Comment (Wovchena) [reply text_generation/causal_lm/cpp/README.md]: Done

File: text_generation/causal_lm/cpp/README.md [hunk 2]
```diff
@@ -10,7 +10,12 @@ A common LLM inference optimisation is introduction of past KV (key/value)-cache
...
+To provide a possibility to implement beam search without accessing model internal state, a stateful LLM converted with `optimum-intel` or [llm_bench](../../../llm_bench/python/) introduces additional 1-dimentional `beam_idx` input. `beam_idx` must contain indices of elements in a batch which are supposed to be selected and evolve during next beam search iteration. There's only one beam when the generation starts. That beam corresponds to initial prompt. `beam_idx` must have values `[0, 0]` to keep initial beam and introduce its copy. Leveraging dynamic batch size allows to change the number of beams dynamicly. `beam_idx` must have `[1]` value to remove zeroth sequence and keep second beam only.
```
Comment (kblaszczak-intel) [inline text_generation/causal_lm/cpp/README.md:13]:
```suggestion
To provide the possibility to implement beam search without accessing model's internal state, a stateful LLM converted with `optimum-intel` or [llm_bench](../../../llm_bench/python/) introduces an additional 1-dimentional `beam_idx` input. `beam_idx` must contain indices of elements in a batch which are intended to be selected and will evolve during the next beam search iteration. There's only one beam when the generation starts. That beam corresponds to the initial prompt. `beam_idx` must have values: `[0, 0]` to keep the initial beam and introduce its copy. The dynamic batch size enables to change the number of beams dynamically. `beam_idx` must have `[1]` as the value to remove zeroth sequence and keep the second beam only.
```
additional question: do you prefer "indices" or "indexes?" Both are correct but indexes is easier to read.

File: text_generation/causal_lm/cpp/README.md [hunk 3]
```diff
@@ -10,7 +10,12 @@ A common LLM inference optimisation is introduction of past KV (key/value)-cache
...
+
+Suppose there are two running beams. To proceed generating both beams at the next iteration, `beam_idx` values must be `[0, 1]`, pointing to batch elements `0` and `1`. To drop the last beam and split the other beam in two, `beam_idx` must be set to `[0, 0]`, this results in utilizing only the part of KV cache corresponding to zeroth element in the batch. The process of selecting appropriate entries in cache is called Cache Reorder.
```
Comment (kblaszczak-intel) [inline text_generation/causal_lm/cpp/README.md:15]:
```suggestion
Assume there are two running beams. To proceed with generating both beams at the next iteration, `beam_idx` values must be `[0, 1]`, pointing to batch elements `0` and `1`. To drop the last beam and split the other beam in two, `beam_idx` must be set to `[0, 0]`. This results in utilizing only the part of KV cache corresponding to the zeroth element in the batch. The process of selecting proper entries in cache is called Cache Reorder.
```

**Review by kblaszczak-intel — APPROVED** [review]
Suggestions added but they are minor and mostly regard word choice.

File: text_generation/causal_lm/cpp/README.md (hunk 2 above)
Comment (Wovchena) [reply text_generation/causal_lm/cpp/README.md]: Moved to indexes

**Pull request merged**